bitbots_vision/bitbots_vision/vision_modules/yoeo_vision_components.py

```diff
@@ -0,0 +1,889 @@
+        self._ball_detector = yoeo_handler.YOEOBallDetector(yoeo)
+        self._debug_image = DebugImageFactory.create(config)
+        self._field_boundary_detector = YOEOFieldBoundaryDetectorFactory.create(config, yoeo)
+
+        self._register_publisher(config)
+        self._config = config
+
+    def _register_publisher(self, new_config: Dict) -> None:
+        self._publisher = ros_utils.create_or_update_publisher(
+            self._node,
+            self._config,
+            new_config,
+            self._publisher,
+            'ROS_ball_msg_topic',
+            BallArray
+        )
+
+    def run(self, image_msg) -> None:
+        """
+        :param image_msg: Image message
+        :type image_msg:  sensor_msgs.msg._image.Image
+        """
+        candidates = self._get_best_ball_candidates()
+        candidates_within_convex_fb = self._filter_for_best_candidates_within_convex_field_boundary(candidates)
+        final_candidates = self._filter_by_candidate_threshold(candidates_within_convex_fb)
+
+        ball_messages = self._create_ball_messages(final_candidates)
+        balls_message = self._create_balls_message(image_msg, ball_messages)
+        self._publish_balls_message(balls_message)
+
+        self._add_candidates_to_debug_image(candidates)
+        self._add_candidates_within_convex_fb_to_debug_image(candidates_within_convex_fb)
+        self._add_final_candidates_to_debug_image(final_candidates)
+
+    def _get_best_ball_candidates(self) -> List[candidate.Candidate]:
+        return self._ball_detector.get_top_candidates(count=self._config['ball_candidate_max_count'])
+
+    def _filter_for_best_candidates_within_convex_field_boundary(self, candidates: List[candidate.Candidate]) \
+            -> List[candidate.Candidate]:
+        return self._field_boundary_detector.candidates_under_convex_field_boundary(
+            candidates,
+            self._config['ball_candidate_field_boundary_y_offset']
+        )
+
+    def _filter_by_candidate_threshold(self, candidates: List[candidate.Candidate]) \
+            -> List[candidate.Candidate]:
+        return candidate.Candidate.rating_threshold(candidates, self._config['ball_candidate_rating_threshold'])
+
+    @staticmethod
+    def _create_ball_messages(candidates: List[candidate.Candidate]):
+        """
+        :rtype: List[soccer_vision_2d_msgs.msg._ball.Ball]  # TODO
+        """
+        return list(map(ros_utils.build_ball_msg, candidates))
+
+    @staticmethod
+    def _create_balls_message(image_msg, ball_messages):
+        """
+        :param image_msg: Image message
+        :type image_msg:  sensor_msgs.msg._image.Image
+        :param ball_messages: List of ball messages
+        :type ball_messages: List[soccer_vision_2d_msgs.msg._ball.Ball]  # TODO
+        :rtype: soccer_vision_2d_msgs.msg._ball_array.BallArray  # TODO
+        """
+        return ros_utils.build_ball_array_msg(image_msg.header, ball_messages)
```

Comment:
This is a pretty simple one liner, I personally don't think these should be wrapped in another function if their behavior is obvious.